rust/arrow/src/array/array.rs

```diff
@@ -1598,6 +1601,229 @@ impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
     }
 }
 
+/// A dictonary array where each element is a single value indexed by an integer key.
+pub struct DictionaryArray {
+    data: ArrayDataRef,
+    values: ArrayRef,
+    raw_values: RawPtrBox<u8>,
+    key_type: DataType,
+}
+
+macro_rules! collect_values {
+    ( $self:ident, $keys:ident, $array_t:ty, $from_fn:ident ) => {{
+        let values = $self
+            .values
+            .as_any()
+            .downcast_ref::<$array_t>()
+            .unwrap()
+            .value_slice(0, $self.values.len());
+        $keys
+            .iter()
+            .enumerate()
+            .map(|(i, &k)| {
+                if $self.is_null(i) {
+                    None
+                } else {
+                    T::$from_fn(values[k])
+                }
+            })
+            .collect()
+    }};
+}
+
+macro_rules! collect_keys {
+    ( $self:ident, $val_t:ty ) => {{
+        $self
+            .key_slice::<$val_t>()
+            .iter()
+            .map(|&k| k as usize)
+            .collect()
+    }};
+}
+
+impl DictionaryArray {
+    /// Returns an reference to the keys of this dictionary.
+    pub fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    /// Returns an reference to the values of this dictionary.
+    pub fn values(&self) -> ArrayRef {
+        self.values.clone()
+    }
+
+    /// Returns a clone of the value type of this dictionary.
+    pub fn value_type(&self) -> DataType {
+        self.values.data().data_type().clone()
+    }
+
+    /// Returns ith value of this dictionary array.
+    pub fn value(&self, i: usize) -> ArrayRef {
+        self.values.slice(self.value_offset(i) as usize, 1)
+    }
+
+    /// Returns the offset for value at index `i`.
+    ///
+    /// Note this doesn't do any bound checking, for performance reason.
+    #[inline]
+    pub fn value_offset(&self, i: usize) -> i32 {
+        self.value_offset_at(i)
+    }
+
+    /// Returns the length for value at index `i`.
+    #[inline]
+    pub fn value_length(&self, _: usize) -> i32 {
```

Comment:
you could omit the `usize` parameter so the API becomes clearer that the same value is always returned. Perhaps `value_length` is not even needed?

Reply:
The value of a dictionary needs some revision. We should probably create a wrapper for an single element ArrayRef that behaves like a single datum.